File: typescript/intersectionOf2ArraysII.ts
// Given two integer arrays nums1 and nums2, return an array of their intersection. Each element in the result must appear as many times as it shows in both arrays and you may return the result in any order.

// O(n^2)
function intersect1(nums1: number[], nums2: number[]): number[] {
    let output: number[] = []
    for(let n1 of nums1){
        for(let j = 0; j < nums2.length; j++){
            if(n1 == nums2[j]){
                output.push(n1)
                delete nums2[j]
                break
            }
        }
    }
    return output
};

// Better O(n+m)
function intersect2(nums1: number[], nums2: number[]): number[] {
    let count = {}
    let result: number[] = []

    for (let num of nums1) {
        count[num] = (count[num] || 0) + 1
    }
    for (let num of nums2) {
        if (count[num] > 0) {
            result.push(num)
            count[num] -= 1
        }
    }
    return result
};

// Better in terms of memory (arrays must be sorted) O(n+m)
function intersect3(nums1: number[], nums2: number[]): number[] {
    let i = 0, j = 0
    let result: number[] = []
    while (i < nums1.length && j < nums2.length) {
        if (nums1[i] == nums2[j]) {
            result.push(nums1[i])
            i++
            j++
        }
        else if (nums1[i] > nums2[j]) {
            j++
        }
        else {
            i++
        }
    }
    return result
}

File: leetCode.com/array/python/intersectionOf2ArraysII.py
# Given two integer arrays nums1 and nums2, return an array of their intersection. Each element in the result must appear as many times as it shows in both arrays and you may return the result in any order.

# O(n^2)
def intersect1(nums1:list, nums2: list):
    result = []
    for i in range(0, len(nums1)):
        for j in range(0, len(nums2)):
            if(nums1[i] == nums2[j]):
                result.append(nums1[i])
                nums2[j] = None
                break
    return result
    
# Better O(n+m)
def intersect2(nums1:list, nums2: list):
    count = {}
    result = []
    for i in range(0, len(nums1)):
        count[nums1[i]] = count.get(nums1[i], 0) + 1
    for i in range(0, len(nums2)):
        if(count.get(nums2[i], 0) > 0):
            result.append(nums2[i])
            count[nums2[i]] -= 1
    return result

# Better in term of memory O(n+m) (arrays must be sorted)
def intersect3(nums1: list, nums2: list):
    result = []
    i = j = 0
    
    while(i < len(nums1) and j < len(nums2)):
        if nums1[i] == nums2[j]:
            result.append(nums1[i])
            i += 1
            j += 1
        elif nums1[i] < nums2[j]:
            i += 1
        else:
            j += 1
    return result